lib/ecto/migrator.ex

```diff
@@ -153,41 +144,65 @@ defmodule Ecto.Migrator do
   end
 
   defp do_down(repo, version, module, opts) do
-    run_maybe_in_transaction(repo, module, fn ->
+    async_migrate_maybe_in_transaction(repo, version, module, :down, opts, fn ->
       attempt(repo, version, module, :forward, :down, :down, opts)
         || attempt(repo, version, module, :backward, :change, :down, opts)
         || {:error, Ecto.MigrationError.exception(
             "#{inspect module} does not implement a `down/0` or `change/0` function")}
     end)
-    |> case do
-      :ok ->
+  end
+
+  defp async_migrate_maybe_in_transaction(repo, version, module, direction, opts, fun) do
+    parent = self()
+    ref = make_ref()
+    task = Task.async(fn -> run_maybe_in_transaction(parent, ref, repo, module, fun) end)
+
+    if migrated_successfully?(ref, task.pid) do
+      try do
+        # The table with schema migrations can only be updated from
+        # the parent process because it has a lock on the table
```

Comment:
There is an alternative approach to this, which is to use advisory locks.

I wrote a patch to Rails a couple of years ago that added this functionality to ActiveRecord. It avoids the problem of locking the schema_migrations table in a separate transaction.

That way you don't need two transactions at all, and can commit to the schema_migrations table directly from the main migration transaction.

Reply:
if we use advisory lock perhaps we can rollback the change that requires `pool_size: 2` for migrations? Ref: https://github.com/elixir-ecto/ecto/pull/2258

Reply:
@samphilipd that's nice to hear. Can you dig a link to that commit/PR please?

Although I would wait before migrating to advisory locks because we need to support the current mechanism anyway for other databases.

Reply:
Both MySQL and Postgres support advisory locks.

@wojtekmach it seems likely that yes, you will be able to do the migration with only one connection if you use advisory locks.

Take a look at [migration.rb in ActiveRecord](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/migration.rb#L1210).

My original PR is [here](https://github.com/rails/rails/pull/22122).

Reply:
Thanks for the info! There is also a discussion [here](https://github.com/rails/rails/issues/22092) and [here](https://github.com/rails/rails/pull/22103).